src/runtime/typemanager.cs

```diff
@@ -422,6 +434,198 @@ internal static IntPtr BasicSubType(string name, IntPtr base_, Type impl)
+                FieldInfo[] fi = type.GetFields();
+                int size = IntPtr.Size;
+                for (int i = 0; i < fi.Length; i++)
+                {
+                    fi[i].SetValue(null, i * size + TypeOffset.ob_size);
+                }
+            }
+
+            public static IntPtr AllocPyTypeSpec(string typename, int obSize, int obFlags, IntPtr slotsPtr)
+            {
+                byte[] ascii = System.Text.Encoding.ASCII.GetBytes(typename);
+
+                //This approach is the same as the one in interop.cs for AllocModuleDef
+                //allocate the size of the struct (which is given by the value of the last
+                //static member and enough space to hold to typename as a char buffer.  The
+                //amount of space needed is the length of the string and the null terminator
+                //char* name member will simply point to the position of the buffer.
+                int size = name_value + ascii.Length + 1;
+                IntPtr specPtr = Marshal.AllocHGlobal(size);
+
+                Marshal.Copy(ascii, 0, specPtr + name_value, ascii.Length);
+                Marshal.WriteIntPtr(specPtr, name, specPtr + name_value);
+                Marshal.WriteByte(specPtr, name_value + ascii.Length, 0);
+
+                Marshal.WriteInt32(specPtr, basicsize, obSize);
+                Marshal.WriteInt32(specPtr, itemsize, 0);
+                Marshal.WriteInt32(specPtr, flags, obFlags);
+                //Util.WriteCLong(specPtr, basicsize, obFlags);
+                //Util.WriteCLong(specPtr, itemsize, 0);
+                //Util.WriteCLong(specPtr, flags, obFlags);
+
+                Marshal.WriteIntPtr(specPtr, slots, slotsPtr);
+                return specPtr;
+            }
+
+            public static int name = 0;
+            public static int basicsize = name + IntPtr.Size;
+            public static int itemsize = basicsize + 4;
+            public static int flags = itemsize + 4;
+            public static int slots = flags + 4;
+
+            public static int name_value = slots + IntPtr.Size;
+        }
+
+
+        internal static IntPtr CreateTypeObject(string name, int obSize, int obFlags, PY_TYPE_SLOT[] type_slots)
+        {
+            //type_slots *must* be terminated by a {0,0} entry.  TODO - Should I check/throw?
+
+            //convert type slot array into intptr
+            int structSize = Marshal.SizeOf(typeof(PY_TYPE_SLOT));
+            GCHandle pinnedArray = GCHandle.Alloc(type_slots, GCHandleType.Pinned);
+
+            //Well, this will leak.  Maybe pinnedArray should be added as a member to managedtype.
+            IntPtr slotsPtr = pinnedArray.AddrOfPinnedObject();
+            //pinnedArray.Free(); //at some point
+
+            //create a type from the spec and return it.
+            IntPtr specPtr = PyTypeSpecOffset.AllocPyTypeSpec(name, obSize, obFlags, slotsPtr);
+            IntPtr typePtr = Runtime.PyType_FromSpec(specPtr);
```

Comment:
@filmor @lostmsu this part is driving me crazy. I fixed some offset issues but this one remains.  In my debugging session ( I figured out how to debug both CPython C code and pythonnet in the same session!)  I observe this:
![image](https://user-images.githubusercontent.com/11414628/89247866-debd8700-d5d3-11ea-97ae-1a4100a6ac0b.png)

which seems like the slotPtr is correctly assigned in the right offset. stepping into python code, I then see this:
![image](https://user-images.githubusercontent.com/11414628/89247908-f85ece80-d5d3-11ea-8434-65ab7d522e1d.png)

All of the values are right except for slots.  somehow it gets a totally different value in slots.  I built/am debugging python 3.8 where the definition of PyType_Spec is:
![image](https://user-images.githubusercontent.com/11414628/89247950-13c9d980-d5d4-11ea-88d9-07fee76824c2.png)

If I manually modify the value of spec->slots in the immediate window before CPython uses spec it seems to be fine.

Any tips would be very welcome here.

Reply:
Well, seeing the CPython code here reminds me, that since they intend this to be a stable API, you don't need offsets at all. Why don't you just make a C# struct with the same layout and use it instead of untyped pointers and `Marshal` stuff?

E.g.
```csharp
[StructLayout(LayoutKind.Sequential)]
unsafe struct PyTypeSpec {
  public IntPtr Name;
  public int BasicSize;
  public int ItemSize;
  public FlagsEnum Flags;
  public PyTypeSlot* Slots;
}
```

Reply:
BTW, `PyType_FromSpec` should return `NewReference`

Reply:
@lostmsu I didn't know I had this option, it looks much easier.  I'll try it.

Reply:
@lostmsu the compiler doesn't allow me to pass ref spec into PyType_FromSpec

Reply:
@koubaa you have to change the PInvoke signature to have `ref` too.

Reply:
@koubaa actually, if `PyType_FromSpec` copies the contents of the spec, you can even change `Name` to be `string` and `Slots` to be an array instead of a  pointer, adding corresponding `[MarshalAs(...)]`. Probably also drop `ref` on the parameter, and change the `TypeSpec` to be a class.

Reply:
@lostmsu I changed it to use a struct and for the PInvoke signature to use ref.  I didn't go as far as MarshalAs but that should be doable too.  Now I no longer see a crash here.  Thanks!